sfxclient/multitokensink.go

```diff
@@ -121,207 +139,331 @@ func newWorker(maxBuffer int, errorHandler func(error) error) *worker {
-				w.buffer = append(w.buffer, data[:remainingBuffer]...)
-				data = data[remainingBuffer:]
-			} else if remainingBuffer >= msgLength {
-				w.buffer = append(w.buffer, data[:msgLength]...)
-				data = data[msgLength:]
-			}
-			if len(w.buffer) >= w.maxBuffer {
-				// if the buffer is full emit the data
-				w.emit(msg.token, w.buffer, nil)
-				w.buffer = w.buffer[:0] // reset the buffer
-			}
+		_ = atomic.AddInt64(&w.stats.TotalDatapointsEmitted, int64(len(datapoints)))
+	}
+}
+
+// processMsg is
+func (w *datapointWorker) processMsg(msg *dpMsg) {
+	for len(msg.data) > 0 {
+		msgLength := len(msg.data)
+		remainingBuffer := (w.batchSize - len(w.buffer))
+		if msgLength > remainingBuffer {
+			msgLength = remainingBuffer
+		}
+		w.buffer = append(w.buffer, msg.data[:msgLength]...)
+		msg.data = msg.data[msgLength:]
+		if len(w.buffer) >= w.batchSize {
+			w.emit(msg.token, w.buffer)
+			w.buffer = w.buffer[:0]
 		}
 	}
-	return
 }
 
 // bufferDatapoints is responsible for batching incoming datapoints into a buffer
-func (w *datapointWorker) bufferFunc(msg *msg) (stop bool) {
+func (w *datapointWorker) bufferFunc(msg *dpMsg) (stop bool) {
 	var lastTokenSeen = msg.token
-	_ = w.processMsg(msg)
+	w.processMsg(msg)
 outer:
-	for len(w.buffer) < w.maxBuffer {
+	for len(w.buffer) < w.batchSize {
 		select {
 		case msg = <-w.input:
 			if msg.token != lastTokenSeen {
 				// if the token changes, then emit what ever is in the buffer before proceeding
-				w.emit(msg.token, w.buffer, nil)
+				w.emit(msg.token, w.buffer)
 				w.buffer = w.buffer[:0]
 				lastTokenSeen = msg.token
 			}
-			if stop = w.processMsg(msg); stop { // will continually read and buffer until there are no more datapoints to read
-				break outer // emit what ever is in the buffer and return true to stop
-			}
+			w.processMsg(msg)
 		default:
 			break outer // emit what ever is in the buffer if there are no more datapoints to read
 		}
 	}
 	// emit the data in the buffer
-	w.emit(msg.token, w.buffer, nil)
+	w.emit(msg.token, w.buffer)
 	w.buffer = w.buffer[:0]
 	return
 }
 
 // newBuffer buffers datapoints and events in the pipeline for the duration specified during Startup
-func (w *datapointWorker) newBuffer() (chan *msg, chan bool) {
-	w.input = make(chan *msg)
-
+func (w *datapointWorker) newBuffer() {
 	go func() {
-		var stop bool
-		for msg := range w.input {
-			if msg == stopMsg {
-				stop = true
-			} else {
-				// process the Datapoint Message
-				stop = w.bufferFunc(msg)
-			}
-			if stop {
-				w.done <- true
+		for {
+			select {
+			case <-w.closing: // check if the worker is in a closing state
+				w.stopped.Set(true)
 				return
+			case msg := <-w.input:
+				// process the Datapoint Message
+				w.bufferFunc(msg)
 			}
 		}
 	}()
-	return w.input, w.done
 }
 
-func newDatapointWorker(maxBuffer int, errorHandler func(error) error) *datapointWorker {
+// Stop will stop the pipeline
+func (w *datapointWorker) Stop(timeout time.Duration) error {
```

Comment:
you're going to have to explain what you expect this to do to me... 

i see you call Stop(30s), it creates a timer for 30s, then the default which will obviously happen since it's not 30s yet, select on something off closing except i see no code that actually ever puts something on the closing chan, so default will be called, which will close the channel and then loop again?